contracts/v0.6/swap/ERC20TokenPeg.sol

```diff
@@ -0,0 +1,116 @@
+
+// SPDX-License-Identifier: MIT
+pragma solidity >0.6.0 <0.8.0;
+
+import "@openzeppelin/contracts/math/SafeMath.sol";
+import "@openzeppelin/contracts/proxy/Initializable.sol";
+import "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import "@openzeppelin/contracts/token/ERC20/SafeERC20.sol";
+import "./ISwap.sol";
+
+/**
+ * @title ERC20TokenPeg
+ * @notice Using this contract users can swap between pegged source and target tokens.
+ * @dev Make sure only to deploy with target ERC20 token having a fixed supply
+ * as this contract assumes a simple 1:1 fixed peg.
+ */
+contract ERC20TokenPeg is ISwap, Initializable {
+  using SafeMath for uint256;
+  using SafeERC20 for IERC20;
+
+  IERC20 private immutable _source;
+  IERC20 private immutable _target;
+
+  /**
+   * @dev Create the ERC20TokenPeg between two tokens
+   *
+   * Requirements:
+   * - `sourceAddr` and `targetAddr` cannot be the zero address.
+   * - `sourceAddr` and `targetAddr` cannot be the same contracts.
+   *
+   * @param sourceAddr The address of source token contract
+   * @param targetAddr The address of target token contract
+   */
+  constructor(
+    address sourceAddr,
+    address targetAddr
+  )
+    public
+  {
+    require(sourceAddr != address(0), "ERC20TokenPeg: sourceAddr is zero address");
+    require(targetAddr != address(0), "ERC20TokenPeg: targetAddr is zero address");
+    require(sourceAddr != targetAddr, "ERC20TokenPeg: sourceAddr == targetAddr");
+    _source = IERC20(sourceAddr);
+    _target = IERC20(targetAddr);
+  }
+
+  /**
+   * @dev Initialize the ERC20TokenPeg, only once!
+   *
+   * Contract will transfer from caller and lock source token in the amount of
+   * the circulating supply of target token. Additionally, the contract will
+   * transfer from caller and lock target tokens in the amount of the
+   * uncirculated supply (callers balance).
+   *
+   * Requirements:
+   * - the contract must have allowance for `caller`'s source tokens of at least
+   * target token circulating supply `amount`.
+   * - the contract must have allowance for `caller`'s target tokens of at least
+   * target token uncirculated supply `amount`.
+   */
+  function init()
```

Comment:
Does this mean that if the contract sees heavy traffic in one direction (for example: source -> target) that the contract might dry up, and the `initializer()` address must manually top up this contract with target tokens?

Reply:
Yes, if the target total supply is less than the source. 
In our case they are equal, but one could use this type of contract to peg a token with lesser supply.

For example, if you deploy an ERC20 token with a total supply of 100 (special edition token), you peg it to any other ERC20 (LINK). Now the market can't deposit more than 100 LINK, as the target supply dries out.

After we created 100 special LINK tokens, we can imagine deploying a liquidity pool between source and target LINK as this will now be a floating peg. The price will move from 1:1 as the supply is limited and if the token is perceived valuable by the market, its price could rise. On the other hand, it should never fall under ~1:1.

We get a similar outcome if we cap the wrapper token, as in `LinkTokenWrapperCapped`.